### PR TITLE
 docs/OSUpgrades.md: New file

### PR DESCRIPTION
**Pull request by cgwalters**
Describe the pivot magic with reference to the installer; we need
developers/testers and users to understand this.

### CODE REVIEW

**Comment by kikisdeliveryservice** [comment]
gah we need to stop running ci against docs.  will review in a bit.

**Comment by cgwalters** [comment]
> gah we need to stop running ci against docs.

The first patch will fix that for `README.md` *after* it merges.

File: docs/OSUpgrades.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+about the early pivot.
+
+First, be sure you've read and understand [openshift/installer overview](https://github.com/openshift/installer/blob/37b99d8c9a3878bac7e8a94b6b0113fad6ffb77a/docs/user/overview.md).
+
+In this example we'll use AWS, but again this is all similar to a situation of
+booting bare metal machines via PXE or VMs in OpenStack.
+
+The openshift-installer starts, and uses the AMI it has pinned as the bootstrap node.
+
+The bootstrap node pulls the release image, which contains a reference to a newer `machine-os-content`.
```
Comment (mhrivnak) [inline docs/OSUpgrades.md]: I've read this a few times, and I'm not clear on what exactly `machine-os-content` is. Is it a reference to some other image that holds an ostree repo? Or is it somehow a part of the release image? I've also read through the CVO documentation and did not find an answer there.

If I look at a release image, where/how would I find `machine-os-content`, and what would I see there?

Comment (cgwalters) [reply docs/OSUpgrades.md]: The release image is a reference to a lot of other images.  Look at e.g. this:

https://github.com/openshift/installer/blob/af56e2db2e2601156a1ebe56269c7deb67f3c725/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L23

As for what's *inside* it, see https://github.com/coreos/coreos-assembler/blob/a358e4475063c0f195508318c7f5c2a20af3b236/src/oscontainer.py

Comment (cgwalters) [reply docs/OSUpgrades.md]: > Is it a reference to some other image that holds an ostree repo?

Yes, exactly.

File: docs/OSUpgrades.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+The installer runs the MCO in "bootstrap" mode to render Ignition configuration.
+As part of that the MCO injects a reference to that updated `machine-os-content`
+into the Ignition in the `/etc/pivot/image-pullspec` directory.
+
+The master machines boot, pulling their Ignition configs from the bootstrap
+node. As part of that `pivot.service` runs on the host (which is
+`Before=kubelet.service`, so before the cluster starts). It detects the
+reference to the updated oscontainer and pulls it, uses it to upgrade, and
```
Comment (mhrivnak) [inline docs/OSUpgrades.md]: Will the oscontainer normally be pulled into a local registry and then made available to each node from there? Or does each node download it from a remote source?

Comment (cgwalters) [reply docs/OSUpgrades.md]: Today, each node downloads it - same as any other container in the payload.  Today we don't do "pull-through" to the local registry.

Comment (mhrivnak) [reply docs/OSUpgrades.md]: Gotcha. Where's the best place for me to submit a feature request? We're working on bare-metal use cases and would prefer to not re-download these images.

Comment (cgwalters) [reply docs/OSUpgrades.md]: This repository is fine.

**Comment by cgwalters** [comment]
:telephone_receiver: ?

File: docs/OSUpgrades.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+In this example we'll use AWS, but again this is all similar to a situation of
+booting bare metal machines via PXE or VMs in OpenStack.
+
+The openshift-installer starts, and uses the AMI it has pinned as the bootstrap node.
+
+The bootstrap node pulls the release image, which contains a reference to a newer `machine-os-content`.
+
+The installer runs the MCO in "bootstrap" mode to render Ignition configuration.
+As part of that the MCO injects a reference to that updated `machine-os-content`
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply docs/OSUpgrades.md]: OK I reworked this a bit more extensively.

**Comment by cgwalters** [comment]
Thanks, good comments!  Updated :new

**Comment by kikisdeliveryservice** [comment]
Looks great @cgwalters 

/lgtm

**Comment by kikisdeliveryservice** [comment]
AWS seems to be having issues right now, I'll make sure this gets in.

**Pull request merged**